### ISSUE TITLE
New OutputFilter: Gatekeeper / ContextManager

### ISSUE THREAD

**Issue by lkraav** [inpsyde/assets#19]
**Is your feature request related to a problem? Please describe.**
There are assets that need to be only conditionally loaded in specific contexts, such as payment gateway scripts not necessary anywhere outside of WooCommerce `is_checkout()`

**Describe the solution you'd like**
New OutputFilter could accept an array of callbacks, or a boolean result to determine whether the asset should be enqueued at all.

This should also work for controlling non-`Inpsyde\Asset` dependencies. We should be able to initialize any externally enqueued handle to an `Inpsyde\Asset` object, then apply any OutputFilters as usual.

**Describe alternatives you've considered**
Not sure, are there any?

**Additional context**
I'm looking to better control my own and various 3rd party assets enqueued to only specific contexts, to optimized asset payloads for increased page load speed.

**Comment by lkraav** [comment]
Can the team get back to me on whether you think this is a suitable addition to this package's functionality? I think I could file a PR then if it seems useful.

cc @gmazzap

**Comment by Chrico** [comment]
Howdy,

you can already enqueue Assets conditionally like following:

```php
<?php
$script = new Script('foo', 'foo.js');
$script->canEnqueue('is_checkout');
// or
$script->canEnqueue(function(): bool {
    return function_exists('is_checkout') && is_checkout();
});
```

Same goes for `Style`. There is a function which allows you to define a callback which returns either `true` (do enqueue)  or `false`  (only register). :)

**Comment by lkraav** [comment]
After some source digging, it seems like https://github.com/inpsyde/assets/blob/master/src/BaseAsset.php could be extended to something like `3rdPartyAsset`, where constructor requires only registered handle, and fills object data from existing WP asset object global database.

**Comment by Chrico** [comment]
Ah now i unterstand. Currently we do not connect to anything which is "outside" of the `AssetManager`. It's probably hard to keep that synced, because you can at nearly every point register/enqueue a new Script/Style.

**Comment by lkraav** [comment]
>  It's probably hard to keep that synced, because you can at nearly every point register/enqueue a new Script/Style.

True, but I'll explore some and see if I can come up with something sensible - as I could use a systemic solution to this anyway, and inpsyde/assets seems to be the highest quality asset mgmt solution currently available in the ecosystem.

**Comment by lkraav** [comment]
> Currently we do not connect to anything which is "outside" of the AssetManager. It's probably hard to keep that synced, because you can at nearly every point register/enqueue a new Script/Style.

We would only take responsibility for things registered by the time of overtake operation.
